File: ctftime-solutions/mixto.py
# Mixto lite lib for python3

from typing import Dict, List, Union
from urllib.request import Request, urlopen
from urllib.parse import urlencode, urljoin
from urllib.error import HTTPError
from pathlib import Path
from os import getenv
import json

MIXTO_ENTRY_ID = getenv("MIXTO_ENTRY_ID")
MIXTO_HOST = getenv("MIXTO_HOST")
MIXTO_API_KEY = getenv("MIXTO_API_KEY")


class MissingRequired(Exception):
    """Missing params"""

    pass


class BadResponse(Exception):
    """Bad response from Mixto API"""

    pass


class MixtoLite:
    def __init__(
        self, host: Union[str, None] = None, api_key: Union[str, None] = None
    ) -> None:
        super().__init__()
        self.host = host
        self.api_key = api_key
        self.workspace_id = None
        self.status = 0
        self.commit_type = "tool"

        # if envars are set, always use those values
        if MIXTO_HOST is not None and self.host is None:
            self.host = MIXTO_HOST
        if MIXTO_API_KEY is not None and self.api_key is None:
            self.api_key = MIXTO_API_KEY

        # if host or apikey is not available, read config file
        if self.host == None or self.api_key == None:
            try:
                conf_path = str(Path().home() / ".mixto.json")
                with open(conf_path) as f:
                    j = json.loads(f.read())
                    self.host = j["host"]
                    self.api_key = j["api_key"]
                    self.workspace_id = j["workspace_id"]
            except:
                print("Cannot read mixto config file")
                raise

    def MakeRequest(
        self,
        method: str,
        uri: str,
        body: dict = {},
        query: dict = {},
        isJSON: bool = True,
    ):
        """Generic method helpful in extending this lib for other Mixto
        API calls. Refer to Mixto docs for all available API endpoints.

        Args:
            method (str): Request method
            uri (str): Mixto URI.
            body (dict, optional): Body. Defaults to {}.
            query (dict, optional): Query params. Defaults to {}.
            isJSON (bool, optional): If the response is of type JSON. Defaults to True.

        Raises:
            BadResponse: [description]
            BadResponse: [description]

        Returns:
            [type]: [description]
        """
        url = urljoin(str(self.host), uri)
        q = ""
        if query:
            q = "?" + urlencode(query)
        req = Request(
            method=method.upper(),
            url=url + q,
            data=json.dumps(body).encode(),
            headers={
                "x-api-key": self.api_key,
                "user-agent": "mixto-lite-py",
            },
        )
        if body:
            req.add_header("Content-Type", "application/json")
        try:
            res = urlopen(req)
            body = res.read().decode()
            self.status = res.getcode()
            if self.status > 300:
                raise BadResponse(self.status, res)
            if isJSON:
                return json.loads(str(body))
            else:
                return body
        except HTTPError as e:
            raise BadResponse(e.code, e.read())

    def AddCommit(
        self, data: str, entry_id: str = None, title: str = "", optional: dict = {}
    ):
        """Add/commit data to an entry. This is the primary functionality of
        an integration

        Args:
            data (str): Data to add
            entry_id (str, optional): Entry ID. Will use MIXTO_ENTRY_ID as primary. Defaults to None.
            title (str, optional): Title for commit. Defaults to "Untitled".
            optional (dict, optional): Optional dict to add to request body.

        Raises:
            MissingRequired: If entry id is missing

        Returns:
            dict: Commit added response
        """
        if MIXTO_ENTRY_ID is None and entry_id is None:
            raise MissingRequired("Entry id is missing")

        e_id = MIXTO_ENTRY_ID if MIXTO_ENTRY_ID else entry_id
        body = {
            "data": data,
            "workspace_id": self.workspace_id,
            "entry_id": e_id,
            "commit_type": self.commit_type,
            "title": title,
        }

        if len(optional) > 0:
            body = body | optional

        r = self.MakeRequest(
            "POST",
            "/api/v1/commit",
            body,
        )
        return r

    def GetWorkspaces(self) -> List[Dict[str, str]]:
        """Get all workspaces information and stats

        Returns:
            List[Dict[str, str]]: Array of workspace items
        """
        return self.MakeRequest("GET", "/api/v1/workspace")["data"]

    def GetEntryIDs(self) -> List[Dict[str, str]]:
        """Get all entry ids filtered by the current workspace

        Returns:
            List[Dict[str, str]]: List of entry ids
        """
        # get all entries
        entries = self.MakeRequest(
            "POST",
            "/api/v1/workspace",
            {"workspace_id": self.workspace_id},
        )["data"]["entries"]
        # filter workspaces by current workspace
        return entries

File: mixto-obsidian/mixto.ts
import { existsSync, readFileSync } from "fs";
import { homedir } from "os";
import * as path from "path";
import * as http from "http";
import * as https from "https";

export class MixtoLite {
	host: string | undefined;
	api_key: string | undefined;
	workspace_id: string | undefined;

	/**
	 *Creates an instance of MixtoLite.
	 * @param {string} [host] The host URL.
	 * @param {string} [apiKey] A valid API key
	 * @memberof MixtoLite
	 */
	constructor(host?: string, apiKey?: string) {
		this.host = host;
		this.api_key = apiKey;
		this.workspace_id = undefined;

		if (!this.host) {
			this.host = process.env.MIXTO_HOST;
		}
		if (!this.api_key) {
			this.api_key = process.env.MIXTO_API_KEY;
		}
		// if host and api key is still empty, read config file
		if (!this.host || !this.api_key) {
			const confPath = path.join(homedir(), ".mixto.json");
			if (!existsSync(confPath)) {
				throw new Error("Mixto config file not found");
			}
			const config = JSON.parse(readFileSync(confPath, "utf-8"));
			this.host = config.host;
			this.api_key = config.api_key;
			this.workspace_id = config.workspace_id;
		}
	}

	/**
	 *A generic request handler that allows making requests 
	 to the Mixto API easy.
	 *
	 * @param {string} endpoint The URI. Example: /api/entry/...
	 * @param {http.RequestOptions} options Request options. Needed 
	 * when building a custom request to the API
	 * @param {*} [data] Data object to send as JSON
	 * @param {*} [query] Query params as an object
	 * @returns Thenable string
	 * @memberof MixtoLite
	 */
	MakeRequest(
		endpoint: string,
		options: http.RequestOptions,
		data?: any,
		query?: any
	) {
		return new Promise((resolve, reject) => {
			const url = new URL(endpoint, this.host);
			if (query) {
				Object.keys(query).map((k) =>
					url.searchParams.append(k, query[k])
				);
			}
			options.headers = {
				"x-api-key": this.api_key,
				"content-type": "application/json",
				"user-agent": "mixto-lite-node",
			};
			const bufferData = Buffer.from(data);
			if (data) {
				options.headers["content-length"] = bufferData.length;
			}
			const reqLib = url.protocol === "https:" ? https : http;
			const req = reqLib.request(url.toString(), options, (res) => {
				var body: any[] = [];
				if (res.statusCode) {
					if (res.statusCode < 200 || res.statusCode >= 300) {
						return reject(
							new Error(`Bad response: ${res.statusCode}`)
						);
					}
				}
				res.on("data", (c) => body.push(c));
				res.on("end", () => {
					resolve(Buffer.concat(body).toString());
				});
				res.on("error", () =>
					reject(`Bad response: ${res.statusCode}`)
				);
			});
			if (bufferData) {
				req.write(bufferData);
			}
			req.end();
		});
	}

	/**
	 *Get an array of all workspaces, entries and commits. 
	 Helpful when creating builders to select an entry.
	 *
	 * @returns {Promise<Workspace[]>} Thenable array of workspaces
	 * @memberof MixtoLite
	 */
	GetWorkspaces(): Promise<Workspace[]> {
		return this.MakeRequest(
			`/api/v1/workspace`,
			{ method: "GET" },
			null
		).then((d) => {
			return JSON.parse(d as any).data as Workspace[];
		});
	}

	/**
	 *Get all entry ids filtered by current workspace.
	 *
	 * @param {boolean} [includeCommits] Include all commits in response
	 * @returns {Promise<Entry[]>}
	 * @memberof MixtoLite
	 */
	async GetEntryIDs(includeCommits: boolean = false): Promise<Entry[]> {
		let body = JSON.stringify({
			workspace_id: this.workspace_id,
			include_commits: includeCommits,
		});
		return this.MakeRequest(
			`/api/v1/workspace`,
			{ method: "POST" },
			body
		).then((res: any) => {
			return JSON.parse(res).data.entries;
		});
	}

	/**
	 *Add a commit to an entry
	 *
	 * @param {*} data Data to commit
	 * @param {string} [entry_id] A valid entry ID
	 * @param {string} [title] Title for the entry. Optional. Defaults
	 * to untitled
	 * @returns {Promise<Commit>} Thenable Commit object
	 * @memberof MixtoLite
	 */
	AddCommit(data: any, entry_id?: string, title?: string): Promise<Commit> {
		if (!entry_id && !process.env.MIXTO_ENTRY_ID) {
			throw new Error("Entry ID not specified");
		}
		let body = JSON.stringify({
			data: data,
			commit_type: "tool",
			title: title,
			workspace_id: this.workspace_id,
			entry_id: entry_id,
			meta: {},
		});
		return this.MakeRequest(
			`/api/v1/commit`,
			{ method: "POST" },
			body
		).then((d) => {
			return JSON.parse(d as any) as Commit;
		});
	}

	/**
	 * Make a generic graphql request to the mixto server
	 * @param query gql query
	 * @param variables optional variables
	 * @returns {Promise<GQLResponse>} Thenable GQLResponse object
	 */
	GraphQL(query: string, variables?: any): Promise<GQLResponse> {
		let body: any = { query: query };
		if (variables) {
			body["variables"] = variables;
		}
		return this.MakeRequest(
			"/api/v1/gql",
			{ method: "POST" },
			JSON.stringify(body)
		).then((d) => {
			return JSON.parse(d as any) as GQLResponse;
		});
	}
}

export interface Workspace {
	workspace_id: string;
	workspace_name: string;
}

export interface Entry {
	entry_id: string;
	title: string;
	commits?: Commit[];
}

export interface Commit {
	commit_it: string;
	title: string;
	type: string;
}

export interface GQLResponse {
	data?: {
		[key: string]: any;
	};
	error?: any;
}
